woocommerce/woocommerce-ios
# Order Details: add email action to `See Receipt` screen

Closes: #5331 


Apology for the 500+ diffs, the test cases and mail composer code moving contributed to the size 🙇🏻‍♀️ 

### Description


After a successful in-person payment, the merchant can still access the receipt in order details > `See receipt` CTA. Before this PR, there is only one action to print the receipt. This PR adds a new action to email the receipt similar to the end of the IPP flow.

#### Attempt using iOS share sheet

First, I tried the iOS share sheet with `UIActivityViewController` with the hope to offer more share options than just email. But after a few attempts and testing (the best attempt in [this branch](https://github.com/woocommerce/woocommerce-ios/compare/feat/5331-share-receipt-from-order-details)), the receipt HTML only works well in the native mail client and not other services like Gmail and Cloud drive. For a merchant, I think the most common use case would be to email the receipt to a customer after payment collection.

#### Major changes

The implementation took me some time to make a few decisions based on the current architecture. Since the email receipt flow is exactly the same as the end of the IPP flow in `CollectOrderPaymentUseCase`, I extracted the existing mail composer code out of the use case to a new class `CardPresentPaymentReceiptEmailCoordinator`. The new coordinator class handles the navigation from a given view controller (IPP or receipt details screen) to present the native mail composer and tracks events on various actions.

In `ReceiptViewModel`, I updated the data flow by moving the receipt content to an observable since the HTML content generation could be non-trivial and we probably don't want to generate the HTML on every tap of the email button. The view model now also provides the necessary data for the receipt email form in an observable `emailFormData`. Under the hood, the receipt content acts like `CurrentValueSubject` so that the latest value is returned for both the content and email data observables (this behavior is covered in a test case).

To add the new email action to the receipt details screen, a navigation item is added next to the print item if email is supported in `ReceiptViewController`.

Please feel free to suggest any changes on the implementation! 🙂 

### Testing instructions


There are two entry points to the receipt email form:

- After completing payment:  --> all events below should have a `card_reader_model` event prop
  -  Tap on Email receipt: `receipt_email_tapped`
      - Send email: `receipt_email_success`
      - Cancel and save as draft: `receipt_email_success`
      - Cancel and discard draft: `receipt_email_canceled`
      - ❓Make the email fail: `receipt_email_failed` (this is hard to repro)
  - After the email form is complete, it should go back to the orders tab or order details
- From an order with a saved receipt:  --> all events below should **not** have a `card_reader_model` event prop
  - Tap See Receipt: `receipt_view_tapped`
    - Tap on Print receipt: `receipt_print_tapped `
        - Print receipt: `receipt_print_success`
        - Cancel printing: `receipt_print_canceled`
        - ❓Make printing fail: `receipt_print_failed` (this is hard to repro)

### Screenshots


You can notice that a new "email" icon is added to the navigation bar:

<img src="https://user-images.githubusercontent.com/1945542/168023265-8c187b5c-6a36-400e-b666-5df5e6b05e95.PNG" width="300" />


---
